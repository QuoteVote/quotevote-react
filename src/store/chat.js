/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit'

const chat = createSlice({
  name: 'chat',
  initialState: {
    submitting: false,
    selectedRoom: null,
  },
  reducers: {
    CHAT_SUBMITTING: (state, action) => {
      state.submitting = action.payload
    },
    SELECTED_CHAT_ROOM: (state, action) => {
      state.selectedRoom = action.payload
    },
  },
})

export const { CHAT_SUBMITTING, SELECTED_CHAT_ROOM } = chat.actions

export default chat.reducer
