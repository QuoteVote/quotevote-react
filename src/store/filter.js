/* eslint-disable no-param-reassign */
import { createSlice } from '@reduxjs/toolkit'

const uiSlice = createSlice({
  name: 'filter',
  initialState: {
    filter: {
      visibility: false,
      value: ['POSTED'],
    },
    date: {
      visibility: false,
      value: '',
    },
    search: {
      visibility: false,
      value: '',
    },
  },
  reducers: {
    FILTER_VISIBILITY: (state, action) => {
      state.filter.visibility = action.payload
    },
    FILTER_VALUE: (state, action) => {
      state.filter.value = action.payload
    },
    DATE_VISIBILITY: (state, action) => {
      state.date.visibility = action.payload
    },
    DATE_VALUE: (state, action) => {
      state.date.value = action.payload
    },
    SEARCH_VISIBILITY: (state, action) => {
      state.search.visibility = action.payload
    },
    SEARCH_VALUE: (state, action) => {
      state.search.value = action.payload
    },
  },
})

export const {
  FILTER_VISIBILITY,
  FILTER_VALUE,
  DATE_VISIBILITY,
  DATE_VALUE,
  SEARCH_VISIBILITY,
  SEARCH_VALUE,
} = uiSlice.actions

export default uiSlice.reducer
