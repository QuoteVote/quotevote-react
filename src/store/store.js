import { combineReducers } from 'redux'
import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit'
import postReducer from 'store/post'
import userReducer from 'store/user'

const rootReducer = combineReducers({
  user: userReducer,
  post: postReducer,
})

// Add middleware the end of the array
const middleware = [...getDefaultMiddleware()]

const devMode = process.env.NODE_ENV === 'development'

export default configureStore({
  reducer: rootReducer,
  middleware,
  devTools: devMode,
})
