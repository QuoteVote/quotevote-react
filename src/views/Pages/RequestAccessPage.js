import React, { useState } from 'react'

// @material-ui/core components
import { makeStyles } from '@material-ui/core/styles'

import { Formik } from 'formik'

// login method
import { tokenValidator } from 'store/user'

import { useDispatch } from 'react-redux'
import { useHistory } from 'react-router-dom'
import styles from 'assets/jss/material-dashboard-pro-react/views/loginPageStyle'

import PlansPage from 'components/RequestAccess/Plans'
import BusinessForm from 'components/RequestAccess/Business'
import PersonalForm from 'components/RequestAccess/Personal'

const useStyles = makeStyles(styles)

export default function RequestAccessPage() {
  const classes = useStyles()
  const dispatch = useDispatch()
  const history = useHistory()
  const [selectedPlan, setSelectedPlan] = React.useState(null)
  const [request, setRequest] = useState(null)
  const [isContinued, setContinued] = useState(false)
  const [requestSubmitted, setRequestSubmitted] = useState({ personal: false, business: false })
  const initialValues = {
    firstName: '',
    lastName: '',
    company: '',
    email: '',
    creditCard: {
      cardNumber: '',
      expiry: '',
      cvc: '',
    },
    cost: 0,
  }

  const renderForm = (formikProps) => {
    if (selectedPlan === 'personal') {
      return <PersonalForm requestSubmitted={requestSubmitted} isContinued={isContinued} setContinued={setContinued} {...formikProps} />
    }
    return <BusinessForm requestSubmitted={requestSubmitted} isContinued={isContinued} setContinued={setContinued} {...formikProps} />
  }

  // TODO: Abstract validation into custom hook
  React.useEffect(() => {
    if (tokenValidator(dispatch)) history.push('/hhsb/Home')
  // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <div className={classes.container}>
      {!selectedPlan || !request ? (
        <PlansPage selectedPlan={selectedPlan} onPlanSelect={setSelectedPlan} setRequest={setRequest} />
      ) : (
        <Formik
          enableReinitialize
          render={(props) => renderForm(props)}
          initialValues={initialValues}
          // validationSchema={}
          onSubmit={() => {
            setRequestSubmitted({ ...requestSubmitted, [`${selectedPlan}`]: true })
          }}
        />
      )}
    </div>
  )
}
