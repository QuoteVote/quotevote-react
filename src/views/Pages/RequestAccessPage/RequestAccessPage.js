import React, { useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import { tokenValidator } from 'store/user'
import { useDispatch, useSelector } from 'react-redux'
import { useHistory } from 'react-router-dom'
import { useForm } from 'react-hook-form'
import { useApolloClient, useMutation } from '@apollo/react-hooks'
import { SET_SELECTED_PLAN } from 'store/ui'
import styles from 'assets/jss/material-dashboard-pro-react/views/loginPageStyle'

import PlansPage from 'components/RequestAccess/Plans/Plans'
import PersonalForm from 'components/RequestAccess/PersonalForm/PersonalForm'
import BusinessForm from 'components/RequestAccess/BusinessForm/BusinessForm'
import { REQUEST_USER_ACCESS_MUTATION } from 'graphql/mutations'
import { GET_CHECK_DUPLICATE_EMAIL } from 'graphql/query'

const useStyles = makeStyles(styles)

export default function RequestAccessPage() {
  const classes = useStyles()
  const dispatch = useDispatch()
  const history = useHistory()
  const selectedPlan = useSelector((state) => state.ui.selectedPlan)
  const [request, setRequest] = useState(null)

  const defaultValues = {
    firstName: '',
    lastName: '',
    companyName: '',
    email: '',
  }
  const [userDetails, setUserDetails] = useState(defaultValues)
  const {
    register, errors, getValues, handleSubmit, setError,
  } = useForm({ defaultValues })
  const [requestInviteSuccessful, setRequestInviteSuccessful] = useState(false)
  const [isContinued, setContinued] = useState(false)
  const [errorMessage, setErrorMessage] = useState(null)

  const cardDefaultValues = {
    cardNumber: '',
    expiry: '',
    cvv: '',
    cost: selectedPlan === 'business' ? 10 : 0,
  }
  const [cardDetails, setCardDetails] = useState(cardDefaultValues)

  const client = useApolloClient()

  const setSelectedPlan = (planType) => {
    dispatch(SET_SELECTED_PLAN(planType))
  }

  const onContinue = async () => {
    const newUserDetails = getValues()
    const { data } = await client.query({
      query: GET_CHECK_DUPLICATE_EMAIL,
      variables: { email: newUserDetails.email },
      fetchPolicy: 'network-only',
    })
    const hasDuplicateEmail = data && data.checkDuplicateEmail.length
    if (hasDuplicateEmail) {
      setError('email', {
        type: 'manual',
        message: 'Email already exist!',
      })
    }
    if (!hasDuplicateEmail && !Object.keys(errors).length) { // if there are no errors proceed to card number form
      setUserDetails(newUserDetails)
      setContinued(true)
    }
  }

  const [requestUserAccess, { data, error, loading }] = useMutation(REQUEST_USER_ACCESS_MUTATION)
  const onSubmit = async () => {
    try {
      // eslint-disable-next-line no-console
      console.log({ userDetails, cardDetails })
      const { fullName, ...otherUserDetails } = userDetails
      const requestUserAccessInput = {
        ...otherUserDetails,
        firstName: fullName || otherUserDetails.firstName,
      }
      const { cardNumber, expiry, cvv } = cardDetails
      if (cardNumber && expiry && cvv) {
        const expirySplit = expiry.split('/')
        requestUserAccessInput.cardDetails = {
          number: cardNumber.replace(/\s+/g, ''),
          exp_month: expirySplit[0].replace(/\s+/g, ''),
          exp_year: `20${expirySplit[1]}`.replace(/\s+/g, ''),
          cvc: cvv.replace(/\s+/g, ''),
        }
      }
      await requestUserAccess({ variables: { requestUserAccessInput } })
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log('onSubmit', e)
    }
  }

  React.useEffect(() => {
    if (error) {
      setErrorMessage(error.toString())
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [error])

  React.useEffect(() => {
    if (data) {
      setRequestInviteSuccessful(true)
    }
  }, [data])

  const renderForm = () => {
    if (selectedPlan === 'personal') {
      return (
        <PersonalForm
          setCardDetails={setCardDetails}
          cardDetails={cardDetails}
          isContinued={isContinued}
          onSubmit={onSubmit}
          errors={errors}
          handleSubmit={handleSubmit}
          onContinue={onContinue}
          setContinued={setContinued}
          register={register}
          requestInviteSuccessful={requestInviteSuccessful}
          errorMessage={errorMessage}
          loading={loading}
        />
      )
    }
    return (
      <BusinessForm
        setCardDetails={setCardDetails}
        cardDetails={cardDetails}
        isContinued={isContinued}
        onSubmit={onSubmit}
        errors={errors}
        handleSubmit={handleSubmit}
        onContinue={onContinue}
        setContinued={setContinued}
        register={register}
        requestInviteSuccessful={requestInviteSuccessful}
        errorMessage={errorMessage}
        loading={loading}
      />
    )
  }

  // TODO: Abstract validation into custom hook
  React.useEffect(() => {
    if (tokenValidator(dispatch)) history.push('/hhsb/Home')
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <div className={classes.container}>
      {!selectedPlan || !request ? (
        <PlansPage
          selectedPlan={selectedPlan}
          onPlanSelect={setSelectedPlan}
          setRequest={setRequest}
          setCardDetails={setCardDetails}
          cardDetails={cardDetails}
        />
      ) : renderForm()}
    </div>
  )
}
