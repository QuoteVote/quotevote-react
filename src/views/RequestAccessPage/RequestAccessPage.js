import React, { useState } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import { tokenValidator } from 'store/user'
import { useDispatch } from 'react-redux'
import { useHistory } from 'react-router-dom'
import { useForm } from 'react-hook-form'
import { useApolloClient, useMutation } from '@apollo/react-hooks'
import styles from 'assets/jss/material-dashboard-pro-react/views/landingPageStyle'

import { REQUEST_USER_ACCESS_MUTATION } from 'graphql/mutations'
import { GET_CHECK_DUPLICATE_EMAIL } from 'graphql/query'

import Grid from '@material-ui/core/Grid'
import Input from '@material-ui/core/Input'
import Button from '../../mui-pro/CustomButtons/Button'

const useStyles = makeStyles(styles)

export default function RequestAccessPage() {
  const classes = useStyles()
  const dispatch = useDispatch()
  const history = useHistory()


  const [userDetails, setUserDetails] = useState('')
  const {
    errors, getValues, setError,
  } = useForm({ userDetails })

  const client = useApolloClient()

  const onContinue = async () => {
    const newUserDetails = getValues()
    const { data } = await client.query({
      query: GET_CHECK_DUPLICATE_EMAIL,
      variables: { email: newUserDetails.email },
      fetchPolicy: 'network-only',
    })
    const hasDuplicateEmail = data && data.checkDuplicateEmail.length
    if (hasDuplicateEmail) {
      setError('email', {
        type: 'manual',
        message: 'Email already exists!',
      })
    }
    if (!hasDuplicateEmail && !Object.keys(errors).length) { // if there are no errors proceed to card number form
      setUserDetails(newUserDetails)
      setContinued(true)
    }
  }

  const [requestUserAccess, { data, error, loading }] = useMutation(REQUEST_USER_ACCESS_MUTATION)
  const onSubmit = async () => {
    try {
      // eslint-disable-next-line no-console
      const requestUserAccessInput = {
        email: userDetails
      }
      await requestUserAccess({ variables: { requestUserAccessInput } })
    } catch (e) {
      // eslint-disable-next-line no-console
      console.log('onSubmit', e)
    }
  }

  React.useEffect(() => {
    if (error) {
      setErrorMessage(error.toString())
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [error])

  React.useEffect(() => {
    if (data) {
      setRequestInviteSuccessful(true)
    }
  }, [data])

  // TODO: Abstract validation into custom hook
  React.useEffect(() => {
    if (tokenValidator(dispatch)) history.push('/hhsb/Home')
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  return (
    <div className={classes.container}>
      <Grid
        container
        display="flex"
        justify="center"
        alignItems="center"
        className={classes.inputContainer}
      >
        <Input
          disableUnderline
          placeholder="Enter Email"
          className={classes.input}
          onChange={(event) => setUserDetails(event.target.value)}
        />
        <Button className={classes.requestAccessBtn} onClick={() => onSubmit()}>Request Invite</Button>
      </Grid>
    </div>
  )
}
