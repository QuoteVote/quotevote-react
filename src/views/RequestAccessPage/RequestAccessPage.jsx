import { useState, useEffect } from 'react'
import { makeStyles } from '@material-ui/core/styles'
import { tokenValidator } from 'store/user'
import { useDispatch } from 'react-redux'
import { useHistory } from 'react-router-dom'
import { useForm } from 'react-hook-form'
import { useApolloClient, useMutation } from '@apollo/react-hooks'
import PersonalForm from 'components/RequestAccess/PersonalForm/PersonalForm'
import styles from 'assets/jss/material-dashboard-pro-react/views/landingPageStyle'

import { REQUEST_USER_ACCESS_MUTATION } from '@/graphql/mutations'
import { GET_CHECK_DUPLICATE_EMAIL } from '@/graphql/query'

import Grid from '@material-ui/core/Grid'
import Input from '@material-ui/core/Input'
import { Typography } from '@material-ui/core'

import Button from '../../mui-pro/CustomButtons/Button'

const useStyles = makeStyles(styles)

export default function RequestAccessPage() {
  const classes = useStyles()
  const dispatch = useDispatch()
  const history = useHistory()

  const [userDetails, setUserDetails] = useState('')
  const [errorMessage, setErrorMessage] = useState()
  const [requestInviteSuccessful, setRequestInviteSuccessful] = useState(false)
  const {
    errors,
  } = useForm({ userDetails })

  const client = useApolloClient()

  const [requestUserAccess] = useMutation(REQUEST_USER_ACCESS_MUTATION)
  const onSubmit = async () => {
    const pattern = new RegExp(/^(("[\w-+\s]+")|([\w-+]+(?:\.[\w-+]+)*)|("[\w-+\s]+")([\w-+]+(?:\.[\w-+]+)*))(@((?:[\w-+]+\.)*\w[\w-+]{0,66})\.([a-z]{2,6}(?:\.[a-z]{2})?)$)|(@\[?((25[0-5]\.|2[0-4][\d]\.|1[\d]{2}\.|[\d]{1,2}\.))((25[0-5]|2[0-4][\d]|1[\d]{2}|[\d]{1,2})\.){2}(25[0-5]|2[0-4][\d]|1[\d]{2}|[\d]{1,2})\]?$)/i)
    const isValidEmail = pattern.test(userDetails)
    if (!isValidEmail) {
      setErrorMessage('This is not a valid email address')
    } else {
      const checkDuplicate = await client.query({
        query: GET_CHECK_DUPLICATE_EMAIL,
        variables: { email: userDetails },
        fetchPolicy: 'network-only',
      })
      const hasDuplicateEmail = checkDuplicate && checkDuplicate.data.checkDuplicateEmail.length
      if (hasDuplicateEmail) {
        setErrorMessage('This email already exists')
      } else if (!hasDuplicateEmail && !Object.keys(errors).length) {
        try {
          // eslint-disable-next-line no-console
          const requestUserAccessInput = {
            email: userDetails,
          }
          await requestUserAccess({ variables: { requestUserAccessInput } })
          setRequestInviteSuccessful(true)
        } catch (e) {
          if (e.message.includes('email: Path `email` is required.')) {
            setErrorMessage('Email is required')
          }
        }
      }
    }
  }

  // TODO: Abstract validation into custom hook
  useEffect(() => {
    if (tokenValidator(dispatch)) history.push('/Home')
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [])

  if (requestInviteSuccessful) {
    return (
      <PersonalForm
        requestInviteSuccessful={requestInviteSuccessful}
      />
    )
  }

  const duplicate = (
    <Typography>{errorMessage}</Typography>
  )

  return (
    <div className={classes.container}>
      <Grid
        container
        display="flex"
        justify="center"
        alignItems="center"
        className={classes.inputContainer}
      >
        <Input
          disableUnderline
          placeholder="Enter Email"
          className={classes.input}
          onChange={(event) => setUserDetails(event.target.value)}
        />
        <Button className={classes.requestAccessBtn} onClick={() => onSubmit()}>Request Invite</Button>
        {duplicate}
      </Grid>
    </div>
  )
}
