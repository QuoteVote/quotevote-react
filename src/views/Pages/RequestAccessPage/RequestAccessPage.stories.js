import React from 'react'
import RequestAccessPage from './RequestAccessPage'

export default {
  component: RequestAccessPage,
  title: 'Request Access',
}

export const Base = () => <RequestAccessPage />
