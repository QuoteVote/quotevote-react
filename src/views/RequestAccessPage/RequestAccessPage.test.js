/* eslint-disable no-undef */
/* eslint-disable react/jsx-no-undef */
import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'
import RequestAccessPage from './RequestAccessPage'
import store from '../../store/store'

const RequestAccessPageWrapper = () => (
  <ApolloProvider client={client}>
    <Provider store={store}>
      <RequestAccessPage />
    </Provider>
  </ApolloProvider>
)

describe('Request Access Page test -', () => {
  it('renders correctly', () => {
    const { container } = render(<RequestAccessPageWrapper />)
    expect(container.firstChild).toMatchSnapshot()
  })
})
