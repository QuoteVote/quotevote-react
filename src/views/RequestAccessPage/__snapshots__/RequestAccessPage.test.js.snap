// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Request Access Page test - renders correctly 1`] = `
<div
  class="makeStyles-container-1"
>
  <div
    class="MuiGrid-root makeStyles-inputContainer-30 MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
    display="flex"
  >
    <div
      class="MuiInputBase-root MuiInput-root makeStyles-input-31"
    >
      <input
        class="MuiInputBase-input MuiInput-input"
        placeholder="Enter Email"
        type="text"
        value=""
      />
    </div>
    <button
      class="MuiButtonBase-root MuiButton-root MuiButton-text makeStyles-button-37 makeStyles-requestAccessBtn-32"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label"
      >
        Request Invite
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div>
      <p
        class="MuiTypography-root MuiTypography-body1"
      />
    </div>
  </div>
</div>
`;
